README.md
# Hu-CLI
Hu-CLI 是一个基于 rollup 的打包工具

<br>
<br>

## 安装

```bash
  > npm install @moomfe/hu-cli
```

<br>

## 指令

```bash
  // 打包构建指令, 会自动对代码进行压缩处理
  > hu build

  // 监听模式
  > hu watch
```

<br>

## 配置
```js
  // hu.config.js
  //  - 此配置表为默认配置
  module.exports = {

    // 打包模式
    //  - 开发环境: 'development' || false
    //  - 生产环境: 'production' || true
    //     - 生产环境会自动调用代码压缩工具对代码进行压缩, 速度较慢
    mode: 'development',

    // 打包入口文件
    input: 'index.js',
    // 打包输出文件
    output: 'index.js',

    // 打包入口目录
    inputDir: './src',
    // 打包输出目录
    outputDir: './dist',

    // 静态资源目录, 相对于 `outputDir` 目录
    assetsDir: '',

    // 输出的文件类型 ( amd, cjs, system, esm, iife, umd )
    format: 'iife',
    // 生成 UMD 和 IIFE 模块的名字
    name: '',

    // 在打包好的文件最顶部插入一段内容
    banner: '',

    // 将代码中指定文本进行替换 ( 对象形式 )
    replace: {
      // 'process.env.NODE_ENV': JSON.stringify('production')
    },
    // 将代码中指定文本进行替换 ( 数组形式 )
    //  - 支持传入正则
    replace: [
      // { from: 'process.env.NODE_ENV', to: JSON.stringify('production') },
      // { from: /aaa(?=-)/, to: 'AAA' }
    ],
    
    // 将代码中指定文本进行替换 - 在写入文件前进行替换
    // 规则与 "replace" 相同
    bundleReplace: {},

    // 外部依赖
    //  - 用于定义项目中的外部依赖
    //  - 完整选项请查看 ( https://github.com/MoomFE/Hu-CLI/blob/master/scripts/options/externals/index.md )
    externals: {},

    /**
     * 提供给 rollup 的插件列表
     *  - 可以自行安装一些其他的 rollup 插件
     * @param config 被解析后的 Hu-CLI 配置
     */
    plugins: ( config ) => {
      // 需要通过这样一个工厂函数对 rollup 插件进行返回
      return [];
    },

    // 传递给内置插件选项的配置
    pluginOptions: {
      // 传递给 banner 插件的配置 ( 以下是默认情况下的配置 )
      banner: {
        // 需要在打包后插入 banner 的文件类型
        extensions: [ '.js', '.css' ],
        // 当前 banner 是否需要自动转换为注释插入
        isComment: true
      },
      // 传递给 rollup-plugin-terser 插件的配置 ( 以下是默认情况下的配置 )
      //  - 完整选项请查看插件官网 ( https://github.com/TrySound/rollup-plugin-terser )
      terser: {
        numWorkers: 1,
        ecma: 6,
        warnings: true,
        compress: { passes: 1, unsafe: true, unsafe_proto: true, unsafe_arrows: true, unsafe_methods: true },
        output: { comments: false }
      },
      // 传递给 @rollup/plugin-commonjs 插件的配置
      //  - 完整选项请查看插件官网 ( https://github.com/rollup/plugins/tree/master/packages/commonjs )
      commonjs: {},
      // 传递给 @rollup/plugin-node-resolve 插件的配置
      //  - 完整选项请查看插件官网 ( https://github.com/rollup/plugins/tree/master/packages/node-resolve )
      nodeResolve: {},
      // 传递给 @moomfe/hu-template-minifier 插件的配置
      //  - 完整选项请查看插件官网 ( https://github.com/MoomFE/Hu-Template-Minifier )
      templateMinifier: {},
      // 传递给 @rollup/plugin-babel 插件的配置 ( 以下是默认情况下的配置 )
      //  - 完整选项请查看插件官网 ( https://github.com/rollup/plugins/tree/master/packages/babel )
      babel: {
        babelrc: false,
        exclude: [/\/node_modules\//],
        presets: [
          ['@babel/preset-env', { useBuiltIns: 'usage', corejs: 3, targets: 'browserslist' }]
        ]
      }
    },

    /**
     * 可直接对传入配置进行修改或返回一个新的配置
     * @param rollupConfig 被解析后的 rollup 配置
     * @param config 被解析后的 Hu-CLI 配置
     */
    configureRollup: ( rollupConfig, config ) => {
      // ...
      // return newRollupConfig;
    },

    // 项目兼容性
    browserslist: [
      'Chrome >= 49',
      'Firefox >= 47',
      'Safari >= 10',
      'Edge >= 14'
    ],

    // 多项目打包
    //  - 在使用了 pipe 选项后, 与 pipe 同级的配置将不再视为一个有效的配置表, 仅用作继承给 pipe 内的打包配置用
    pipe: [
      // 会自动继承上一层级的配置
      // {
      //   input: '',
      //   output: ''
      // }
    ]

  };
```

<br>

## License

Hu-CLI is licensed under a [MIT License](./LICENSE).

CHANGELOG.md
  - 🌟 支持导入 `.css`, `.scss`, `.sass` 类型的文件
      - 导入 CSS 文件时可使用 `toString` 参数导入解析过后的 CSS 字符串: `import 'index.css?toString'`
      - 导入 CSS 文件时可使用 `insert` 参数将 CSS 插入到 DOM 中: `import 'index.css?insert'`
      - 导入 CSS 文件时未使用参数, 构建完成后会将导入的 CSS 文件输出为文件
      - 当 `mode` 选项为生产模式 ( `'production' || true` ) 时, 会自动压缩 CSS 代码
  - 🌟 新增 `init` 指令用于初始化项目
  - 🌟 新增 `bundleReplace` 选项, 用于在写入文件前将代码中指定文本进行替换
  - 🌟 新增 `assetsDir` 选项, 用于配置静态资源目录
  - 💄 更新 `banner` 插件将传入的 `banner` 转为注释的细节, 不再生成多余的空格
  - 💄 始终加载 `@moomfe/hu-template-minifier` 用于模板压缩

# 1.9.1
  - 📅 2020-06-09
  - 💄 更新 `eslint` 规则
  - 💄 更新 `babel` 配置解析规则, `@babel/preset-env` 预设的 `targets` 选项允许使用字符串 `browserslist` 来代表当前配置文件的 `browserslist` 配置

# 1.9.0
  - 📅 2020-06-09
  - 🌟 添加 `babel` 相关类库以处理兼容性
  - 🌟 新增 `pluginOptions.babel` 选项, 用于配置传递给内部使用的 `@rollup/plugin-babel` 插件的配置
  - 💄 更换 `rollup-plugin-commonjs` 为 `@rollup/plugin-commonjs`
  - 💄 更换 `rollup-plugin-node-resolve` 为 `@rollup/plugin-node-resolve`
  - 💄 更换 `rollup-pluginutils` 为 `@rollup/pluginutils`

# 1.8.11
  - 📅 2020-04-01
  - 💄 更新 `eslint` 规则

# 1.8.9 - 1.8.10
  - 📅 2020-03-31
  - 💄 更新 `eslint` 规则
  - 💄 更新依赖库, 均采用目前最新版本

# 1.8.6 - 1.8.8
  - 📅 2020-03-27
  - 💄 更新依赖库, 均采用目前最新版本
  - 💄 内部代码启用 `eslint` 进行格式化及代码检查

# 1.8.5
  - 📅 2020-03-10
  - 💄 更新依赖库, 均采用目前最新版本

# 1.8.4
  - 📅 2019-12-20
  - 💄 优化控制台信息的显示

# 1.8.2 - 1.8.3
  - 📅 2019-12-16
  - 💄 更新依赖库, 解决一些问题

# 1.8.1
  - 📅 2019-12-12
  - 💄 当使用 `-c, --config` 指定了配置文件, 则按照配置文件的层级作为根目录

# 1.8.0
  - 📅 2019-12-12
  - 🌟 使用 `watch` / `build` 指令时可以使用 `-c`, `--config` 指定配置文件
  - ⚠️ 移除 `pluginOptions.json` 选项

# 1.7.0
  - 📅 2019-12-10
  - 🌟 增加对 `.json` 文件的支持
  - 🌟 增加对 `.json5` 文件的支持
  - 💄 选项 `pluginOptions` 支持传入 `null` 从而完全不使用从父级继承的配置
  - 💄 选项 `plugins` 支持传入 `null` 从而完全不使用从父级继承的配置
  - 💄 选项 `externals` 的功能完全重构, 功能更加强大
  - 💄 选项 `format` 的可选值支持添加自定义命名空间, 如: `esm.browser`, `umd.xxx`
  - ⚠️ 移除选项 `format` 的可选值 `esm.browser`

# 1.6.0
  - 📅 2019-11-25
  - 🌟 新增 `externals` 选项, 用于标识外部依赖
  - 💄 <s>选项 `format` 新增可选值 `esm.browser`, 表现和 `esm` 完全一致, 只用作 `externals` 的区分上</s> ( 1.7.0 移除 )

# 1.5.1
  - 📅 2019-11-20
  - 🌟 新增 `pluginOptions.templateMinifier` 选项, 用于配置传递给内部使用的 `@moomfe/hu-template-minifier` 插件的配置
  - 💄 为 `plugins`, `configureRollup` 选项内传入的 `config` 参数添加 `_originConfig` 选项, 为原始配置
  - 💄 为 `plugins`, `configureRollup` 选项内传入的 `config` 参数添加 `_mergedConfig` 选项, 为原始配置项与默认配置合并后的配置

# 1.5.0
  - 📅 2019-11-12
  - 🌟 在开发模式下会使用 `@moomfe/hu-template-minifier` 对模板进行压缩

# 1.4.4
  - 📅 2019-11-05
  - 💄 优化控制台信息的显示
  - 💄 将开发模式下默认使用 `terser` 压缩代码的次数由 3 次改为 1 次
  - 💄 将 `banner` 插件的顺序提前, 防止自行安装的插件在使用 `generateBundle` 选项时取到的代码是不带 `banner` 输出的

# 1.4.3
  - 📅 2019-10-31
  - 🌟 新增 `pluginOptions.terser` 选项, 用于配置传递给内部使用的 `rollup-plugin-terser` 插件的配置
  - 🌟 新增 `pluginOptions.commonjs` 选项, 用于配置传递给内部使用的 `@rollup/plugin-commonjs` 插件的配置
  - 🌟 新增 `pluginOptions.nodeResolve` 选项, 用于配置传递给内部使用的 `@rollup/plugin-node-resolve` 插件的配置

# 1.4.2
  - 📅 2019-10-30
  - 💄 `plugins` 选项会在执行时传入解析后的 Hu-CLI 配置
  - 💄 `configureRollup` 选项会在执行时传入解析后的 Hu-CLI 配置

# 1.4.1
  - 📅 2019-10-30
  - 💄 优化控制台警告信息的显示
  - 💄 优化控制台在一些边界情况下信息的显示
  - 💄 优化在使用 watch 指令时, 错误提示的显示

# 1.4.0
  - 📅 2019-10-29
  - 🌟 新增 `mode` 选项, 用于指定当前配置的打包模式
  - ⚠️ 不再根据执行的指令判断是否对代码进行压缩

# 1.3.0
  - 📅 2019-10-29
  - 🌟 新增 `configureRollup` 方法选项, 可用于对解析后配置的修改

# 1.2.0
  - 📅 2019-10-28
  - ⚠️ 在使用了 pipe 选项后, 与 pipe 同级的配置将不再视为一个有效的配置表, 仅用作继承给 pipe 内的打包配置用
  - 🐞 修复使用打包指令时提示错误导致退出的问题

# 1.1.0
  - 📅 2019-10-28
  - 🌟 新增 `format` 选项, 定义输出的文件类型
  - 🌟 新增 `name` 选项, 定义生成 UMD 和 IIFE 模块的名字

# 1.0.0
  - 📅 2019-10-25
  - 🌟 基本功能已完成

<br>
<hr>
<br>
版本规范

1. 主版本号: 破坏性更新和新特性
2. 次版本号: 向下兼容的功能新增、功能更改、功能优化
3. 修订版本号: 向下兼容的问题修正、一般功能优化
<br>
<hr>
<br>
CHANGELOG 图标规范

- 🌟: 功能新增<br>
- 💄: 功能更改、功能优化<br>
- ⚠️: 与上一版本可能不兼容的功能更改<br>
- 🐞: 问题修正<br>
- 📅: 版本发布日期

scripts/config.js
module.exports = {

  // 打包模式
  //  - 开发环境: 'development' || false
  //  - 生产环境: 'production' || true
  mode: 'development',

  // 打包入口文件
  input: 'index.js',
  // 打包输出文件
  output: 'index.js',

  // 打包入口目录
  inputDir: './src',
  // 打包输出目录
  outputDir: './dist',

  // 静态资源目录, 相对于 `outputDir` 目录
  assetsDir: '',

  // 输出的文件类型 ( amd, cjs, system, esm, iife, umd )
  format: 'iife',
  // 生成 UMD 和 IIFE 模块的名字
  name: '',

  // 在打包好的文件最顶部插入一段内容
  banner: '',

  // 将代码中指定文本进行替换
  replace: {},
  // 将代码中指定文本进行替换 - 在写入文件前进行替换
  bundleReplace: {},

  // 外部依赖
  externals: {},

  // 提供给 rollup 的插件列表
  plugins: (config) => {
    return [];
  },

  // 传递给内置插件选项的配置
  pluginOptions: {
    // 传递给 banner 插件的配置
    banner: {
      extensions: ['.js', '.css'],
      isComment: true
    },
    // 传递给 rollup-plugin-terser 插件的配置
    terser: {
      numWorkers: 1,
      ecma: 6,
      warnings: true,
      compress: { passes: 1, unsafe: true, unsafe_proto: true, unsafe_arrows: true, unsafe_methods: true },
      output: { comments: false }
    },
    // 传递给 @rollup/plugin-commonjs 插件的配置
    commonjs: {},
    // 传递给 @rollup/plugin-node-resolve 插件的配置
    nodeResolve: {},
    // 传递给 @moomfe/hu-template-minifier 插件的配置
    templateMinifier: {},
    // 传递给 @rollup/plugin-babel 插件的配置
    babel: {
      babelrc: false,
      exclude: [/\/node_modules\//],
      presets: [
        ['@babel/preset-env', { useBuiltIns: 'usage', corejs: 3, targets: 'browserslist' }]
      ]
    }
  },

  // 方法会接收被解析的 rollup 配置作为参数, 可直接对传入配置进行修改或返回一个新的配置
  configureRollup: (rollupConfig, config) => {

    // return newRollupConfig;
  },

  // 项目兼容性
  browserslist: [
    'Chrome >= 49',
    'Firefox >= 47',
    'Safari >= 10',
    'Edge >= 14'
  ],

  // 多项目打包
  pipe: [
    // {
    //   input: '',
    //   output: ''
    // }
  ]

};


/* ------------------------------------ ------------------------------------ ------------------------------------ */
/* ------------------------------------ ------------------------------------ ------------------------------------ */
/* ------------------------------------ ------------------------------------ ------------------------------------ */


/**
 * 项目兼容性 ( browserslist )
 * 默认项目兼容性的判定标准为支持以下特性:
 *  - "Proxy object" (https://developer.mozilla.org/zh-CN/docs/Web/JavaScript/Reference/Global_Objects/Proxy)
 *  - "Arrow Functions" (https://developer.mozilla.org/zh-CN/docs/Web/JavaScript/Reference/Functions/Arrow_functions)
 *  - "Default parameters" (https://developer.mozilla.org/zh-CN/docs/Web/JavaScript/Reference/Functions/Default_parameters)
 *  - "Rest Parameters" (https://developer.mozilla.org/zh-CN/docs/Web/JavaScript/Reference/Functions/Rest_parameters)
 *  - "Template Strings" / "Template literals" (https://developer.mozilla.org/zh-CN/docs/Web/JavaScript/Reference/template_strings)
 *  - "Computed property names" (https://developer.mozilla.org/zh-CN/docs/Web/JavaScript/Reference/Operators/Object_initializer#计算属性名)
 *  - "Shorthand property names" (https://developer.mozilla.org/zh-CN/docs/Web/JavaScript/Reference/Operators/Object_initializer#属性定义)
 *  - "Shorthand method names" (https://developer.mozilla.org/zh-CN/docs/Web/JavaScript/Reference/Operators/Object_initializer#方法定义)
 *  - "Spread syntax" (https://developer.mozilla.org/zh-CN/docs/Web/JavaScript/Reference/Operators/Spread_syntax)
 *    - "Spread in array literals"
 *    - "Spread in function calls"
 *  - "Destructuring assignment" (https://developer.mozilla.org/zh-CN/docs/Web/JavaScript/Reference/Operators/Destructuring_assignment)
 *    - "Destructuring assignment"
 *    - "Computed property names"
 *  - "Classes" (https://developer.mozilla.org/zh-CN/docs/Web/JavaScript/Reference/Classes)
 *  - "for...of" (https://developer.mozilla.org/zh-CN/docs/Web/JavaScript/Reference/Statements/for...of)
 *  - "const" (https://developer.mozilla.org/zh-CN/docs/Web/JavaScript/Reference/Statements/const)
 *  - "let" (https://developer.mozilla.org/zh-CN/docs/Web/JavaScript/Reference/Statements/let)
 */
